File: backend/app/schemas/billing.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class CheckoutSessionRequest(BaseModel):
    tier: Literal["pro", "business"]


class CheckoutSessionResponse(BaseModel):
    checkout_url: str


class PortalSessionResponse(BaseModel):
    portal_url: str


class PlanOut(BaseModel):
    tier: Literal["free", "pro", "business"]
    available: bool
    unit_amount: int | None = None
    currency: str | None = None
    interval: str | None = None
    upload_limit_per_day: int | None = None


class SubscriptionOut(BaseModel):
    tier: Literal["free", "pro", "business"]
    status: str
    current_period_end: datetime | None = None
    cancel_at_period_end: bool = False


class InvoiceOut(BaseModel):
    id: str
    amount_paid: int
    currency: str
    status: str | None
    created: datetime
    hosted_invoice_url: str | None
    invoice_pdf: str | None


class PaymentMethodOut(BaseModel):
    id: str
    brand: str
    last4: str
    exp_month: int
    exp_year: int
    is_default: bool

File: frontend/types/billing.ts
export type PlanTier = "free" | "pro" | "business";

export interface PlanOut {
  tier: PlanTier;
  available: boolean;
  unit_amount: number | null;
  currency: string | null;
  interval: string | null;
}

export interface SubscriptionOut {
  tier: PlanTier;
  status: string;
  current_period_end: string | null;
  cancel_at_period_end: boolean;
}

export interface InvoiceOut {
  id: string;
  amount_paid: number;
  currency: string;
  status: string | null;
  created: string;
  hosted_invoice_url: string | null;
  invoice_pdf: string | null;
}

export interface PaymentMethodOut {
  id: string;
  brand: string;
  last4: string;
  exp_month: number;
  exp_year: number;
  is_default: boolean;
}
